resources/arduino_files/telemetry_board/telemetry_board.ino

```diff
@@ -1,259 +1,603 @@
-  //      Example serial input:
-  //           4,1   # Turn fan on
-  //          13,0   # Turn led off
-  while (Serial.available() > 0) {
-    int pin_num = Serial.parseInt();
-    int pin_status = Serial.parseInt();
-
-    switch (pin_num) {
-      case COMP_RELAY:
-        /* The computer shutting itself off:
-         *  - Power down
-         *  - Wait 30 seconds
-         *  - Power up
-         */
-        if (pin_status == 0){
-          turn_pin_off(COMP_RELAY);
-          delay(1000 * 30);
-          turn_pin_on(COMP_RELAY);
+//////////////////////////////////////////////////////////////////////////////
+// Input Handlers: the support collecting the values of various sensors/pins,
+// and then reporting it later.
+
+class DHTHandler {
+  public:
+    DHTHandler(uint8_t pin, uint8_t type) : dht_(pin, type) {}
+
+    void Init() {
+      dht_.begin();
+    }
+
+    void Collect() {
+      // Force readHumidity to actually talk to the device;
+      // otherwise will read at most every 2 seconds, which
+      // is sometimes just a little too far apart.
+      // Note that the underlying read() routine has some big
+      // delays (250ms and 40ms, plus some microsecond scale delays).
+      humidity_ = dht_.readHumidity(/*force=*/true);
+      // readTemperature will use the data collected by
+      // readHumidity, which is just fine.
+      temperature_ = dht_.readTemperature();
+    }
+
+    void Report() {
+      // This is being added to a JSON dictionary, so print a comma
+      // before the quoted name, which is then followed by a colon.
+      Serial.print(", \"humidity\":");
+      Serial.print(humidity_);
+      Serial.print(", \"temp_00\":");
+      Serial.print(temperature_);
+    }
+
+  private:
+    DHT dht_;
+    float humidity_;
+    float temperature_;  // Celcius
+} dht_handler(DHT_PIN, DHTTYPE);
+
+// DallasTemperatureHandler collects temp values from Dallas One-Wire temp sensors.
+template <size_t kMaxSensors>
+class DallasTemperatureHandler {
+  public:
+    DallasTemperatureHandler(OneWire* one_wire) : dt_(one_wire), device_count_(0) {}
+
+    void Init() {
+      dt_.begin();
+      dt_.setWaitForConversion(true);
+
+      uint8_t devices = dt_.getDeviceCount();
+      for (uint8_t device_num = 0; device_num < devices; ++device_num) {
+        if (dt_.getAddress(devices_[device_count_].address, device_num)) {
+          devices_[device_count_].index = device_num;
+          ++device_count_;
+          if (device_count_ >= kMaxSensors) {
+            break;
+          }
         }
-        break;
-      case CAMERAS_RELAY:
-      case FAN_RELAY:
-      case WEATHER_RELAY:
-      case MOUNT_RELAY:
-        if (pin_status == 1) {
-          turn_pin_on(pin_num);
-        } else if (pin_status == 0) {
-          turn_pin_off(pin_num);
-        } else if (pin_status == 9) {
-          toggle_pin(pin_num);
+      }
+      // Sort the devices so we get a consistent order from run to run...
```

Comment:
I'm surprised we need to do this. Is it due to DallasTemperature or the underlying OneWire?

Reply:
I don't now recall if there is an inconsistency in the order from run-to-run, but I assume so since I went to the trouble. Fortunately, I did buy some more of those sensors and can test sometime Real Soon Now :registered:. The main point is that it doesn't make a lot of sense to have 3 temps listed, but not have any idea which is which.